test/runtests.jl

```diff
@@ -239,6 +239,17 @@ include("setup_tests.jl")
                 @test_throws ArgumentError test_hash(x -> x + 1)
             end
 
+            if V >= 4
+                @testset "Nested Any" begin
+                    @test test_hash(Dict{Symbol,Any}(:a => NumberTypeA(1))) !=
+                          test_hash(Dict{Symbol,Any}(:a => NumberTypeB(1)))
+                    @test test_hash(Pair{Symbol,Any}(:a, NumberTypeA(1))) !=
+                          test_hash(Pair{Symbol,Any}(:a, NumberTypeB(1)))
+                    @test test_hash(Pair{Symbol,Any}[:a => NumberTypeA(1)]) !=
+                          test_hash(Pair{Symbol,Any}[:a => NumberTypeB(1)])
+                end
```

Comment:
These makes sense. Tests are done to make sure that the specific types (`NumberTypeA` and `NumberTypeB`) are considered during the hashing process (despite the field to be of type `Any`).